aten/src/ATen/native/mps/kernels/Pooling.h

```diff
@@ -0,0 +1,30 @@
+#pragma once
+
+#ifndef __METAL__
+#include <array>
+using ulong = unsigned long;
+#define _ARRAY_NS std
+#else
+#include <metal_array>
+#define _ARRAY_NS metal
+#endif
+
+// N is the maximum allowed number of dimensions in the input and outputs. The
+// maximum allowed pooling dimensions is N-2, because the input may have up to 2
+// leading dimensions that are not pooled. To support up to 3-D pooling, N=5 is
+// the default.
+template <unsigned N = 5>
+struct PoolingParams {
+  int32_t dims;
+  int32_t pooling_dims;
+  _ARRAY_NS::array<int64_t, N> input_sizes;
```

Comment:
One small thing to note about providing these parameters in a struct is that the buffers in the struct need to be a fixed size at compile time (if I understand correctly), whereas before I made this change, the sizes could be determined at runtime. So this puts a limit on how many dimensions we can support while reusing the same code. But at the moment, pytorch doesn't support anything higher than 3-D pooling with 2 leading dims and I'm guessing there aren't any plans to support any higher dimensionality, so it's probably not really an issue. Of course if higher dims are needed in the future, we can just switch it back to specifying these params directly instead of in a struct

Reply:
Yes, this is correct, but even for total number of dims there is an upper limit on total number of tensor dimensions, which I think is currently 16

Reply:
Oh I didn't realize that, good to know